### PR TITLE
Add EKS cluster name detection

### LINKED ISSUE #26794: [processor/resourcedetection]: Add support for k8s.cluster.name where possible
### Component(s)

processor/k8sattributes

### Is your feature request related to a problem? Please describe.

When using the `k8sattributes` processor, I shouldn't need to declare any attributes manually. I still need to manually specify `k8s.cluster.name` as the processor doesn't support it yet.

There is no standard API to get the cluster name, but GKE and EKS make it easy to detect. We should support this in the processor and provide `k8s.cluster.name` where possible.

### Describe the solution you'd like

For GKE, See: https://github.com/open-telemetry/opentelemetry-js-contrib/blob/deb9aa441dc7d2b0fd5ec11b41c934a1e93134fd/detectors/node/opentelemetry-resource-detector-gcp/src/detectors/GcpDetector.ts#L81 (it gets it from `curl http://metadata/computeMetadata/v1/instance/attributes/cluster-name -H "Metadata-Flavor: Google"`)

For EKS see: https://github.com/open-telemetry/opentelemetry-js-contrib/blob/deb9aa441dc7d2b0fd5ec11b41c934a1e93134fd/detectors/node/opentelemetry-resource-detector-aws/src/detectors/AwsEksDetector.ts#L86 (it gets it from the cert name on `kubernetes.default.svc`, iiuc)

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by crobert-1**
**Description:** <Describe what has changed.>

This enhancement detects the k8s cluster name in EKS. The solution uses EC2 instance tags to determine the cluster name, which means it will only work on EC2 (as noted in documentation updates).

**Link to tracking Issue:** <Issue number if applicable>
Resolves #26794

**Testing:** <Describe what testing was performed and which tests were added.>
Manual testing in an EKS cluster when running on EC2 instances.

**Documentation:** <Describe the documentation added.>
Updated README and relevant metadata

### CODE REVIEW

File: processor/resourcedetectionprocessor/internal/aws/eks/metadata.yaml [hunk 1]
```diff
@@ -10,4 +10,8 @@ resource_attributes:
   cloud.platform:
     description: The cloud.platform
     type: string
+    enabled: true
+  k8s.cluster.name:
+    description: The EKS cluster name. This attribute is currently only available when running on EC2 instances, and requires permission to run the EC2:DescribeInstances action.
+    type: string
     enabled: true
```
Comment (crobert-1) [inline processor/resourcedetectionprocessor/internal/aws/eks/metadata.yaml:17]: This will be enabled by default, let me know if it should be disabled by default.

File: processor/resourcedetectionprocessor/internal/aws/eks/detector.go [hunk 2]
```diff
@@ -74,9 +82,55 @@ func (d *detector) Detect(ctx context.Context) (resource pcommon.Resource, schem
...
+	if err != nil {
+		return pcommon.NewResource(), "", err
+	}
+	d.rb.SetK8sClusterName(clusterName)
+
 	return d.rb.Emit(), conventions.SchemaURL, nil
 }
 
+func (d *detector) getClusterName() (string, error) {
+	cfg, err := config.LoadDefaultConfig(context.TODO())
```
Comment (dashpole) [inline processor/resourcedetectionprocessor/internal/aws/eks/detector.go:95]: should this use the context.Context from Detect?  Same comment for context.TODOs below.

File: processor/resourcedetectionprocessor/internal/aws/eks/detector.go [hunk 3]
```diff
@@ -74,9 +82,55 @@ func (d *detector) Detect(ctx context.Context) (resource pcommon.Resource, schem
...
+	if err != nil {
+		d.logger.Debug("Failed to load default config", zap.Error(err))
```
Comment (dashpole) [inline processor/resourcedetectionprocessor/internal/aws/eks/detector.go:97]: generally, we should log or return the error, not both.  I would expect whatever invokes Detect() to handle the error appropriately (e.g. by logging it).  Applies here and below.

File: processor/resourcedetectionprocessor/internal/aws/eks/detector.go [hunk 4]
```diff
@@ -74,9 +82,55 @@ func (d *detector) Detect(ctx context.Context) (resource pcommon.Resource, schem
 	d.rb.SetCloudProvider(conventions.AttributeCloudProviderAWS)
 	d.rb.SetCloudPlatform(conventions.AttributeCloudPlatformAWSEKS)
 
+	clusterName, err := d.getClusterName()
+	if err != nil {
+		return pcommon.NewResource(), "", err
```
Comment (dashpole) [inline processor/resourcedetectionprocessor/internal/aws/eks/detector.go:87]: Will this be breaking for users that don't have the correct permissions?  Would returning `d.rb.Emit(), conventionss.SchemaURL, err` be better, or would the processor still use an empty resource in that case?

File: processor/resourcedetectionprocessor/internal/aws/eks/detector.go [hunk 5]
```diff
@@ -74,9 +82,55 @@ func (d *detector) Detect(ctx context.Context) (resource pcommon.Resource, schem
...
+		for _, instance := range reservation.Instances {
+			for _, tag := range instance.Tags {
+				if *tag.Key == clusterNameAwsEksTag || *tag.Key == clusterNameEksTag {
+					return *tag.Value
+				}
+			}
+		}
+	}
+
+	d.logger.Debug("Could not find cluster name within EC2 instance tags")
```
Comment (dashpole) [inline processor/resourcedetectionprocessor/internal/aws/eks/detector.go:130]: should this return an error, rather than setting the cluster name to empty?

**Review by dashpole — COMMENTED** [review]
LGTM overall, but someone from AWS should take a look.

File: processor/resourcedetectionprocessor/README.md [hunk 6]
```diff
@@ -133,7 +133,7 @@ We map these environment variables to resource attributes as follows:
 | `HEROKU_RELEASE_VERSION`           | `service.version`                   |
 | `HEROKU_SLUG_COMMIT`               | `heroku.release.commit`             |
 
-For more information, see the [Heroku cloud provider documentation](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/semantic_conventions/cloud_provider/heroku.md) under the [OpenTelemetry specification semantic conventions](https://github.com/open-telemetry/opentelemetry-specification).
+For more information, see the [Heroku cloud provider documentation](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/resource/cloud_provider/heroku.md) under the [OpenTelemetry specification semantic conventions](https://github.com/open-telemetry/semantic-conventions).
```
Comment (crobert-1) [inline processor/resourcedetectionprocessor/README.md:136]:
```suggestion
For more information, see the [Heroku cloud provider documentation](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/resource/cloud-provider/heroku.md) under the [OpenTelemetry specification semantic conventions](https://github.com/open-telemetry/semantic-conventions).
```

File: processor/resourcedetectionprocessor/internal/aws/eks/detector.go (hunk 2 above)
Comment (crobert-1) [reply processor/resourcedetectionprocessor/internal/aws/eks/detector.go:95]: I'm not the most well versed here on why, but experimentally directly using `ctx` is causing `context deadline exceeded`. Is it preferable to use it directly and set a longer timeout than using `context.TODO()`?

File: processor/resourcedetectionprocessor/internal/aws/eks/detector.go (hunk 3 above)
Comment (crobert-1) [reply processor/resourcedetectionprocessor/internal/aws/eks/detector.go:97]: That's fair, I was mostly "inheriting" the style of the failure to detect an EKS environment in `Detect()`. In testing though, the messages are duplicated so I agree with your thought. I'll update.

File: processor/resourcedetectionprocessor/internal/aws/eks/detector.go (hunk 4 above)
Comment (crobert-1) [reply processor/resourcedetectionprocessor/internal/aws/eks/detector.go:87]: Good call. It's better to return as much as possible, especially when the other values here aren't dependent on the cluster name (and we've already successfully determined it's an EKS cluster). I'll update.

File: processor/resourcedetectionprocessor/internal/aws/eks/detector.go (hunk 5 above)
Comment (crobert-1) [reply processor/resourcedetectionprocessor/internal/aws/eks/detector.go:130]: I've decided to just return an empty string, and then create an error within `getClusterName` if it's empty.

File: processor/resourcedetectionprocessor/internal/aws/eks/detector.go [hunk 7]
```diff
@@ -28,10 +32,15 @@ const (
 	kubernetesServiceHostEnvVar = "KUBERNETES_SERVICE_HOST"
 	authConfigmapNS             = "kube-system"
 	authConfigmapName           = "aws-auth"
+
+	clusterNameAwsEksTag = "aws:eks:cluster-name"
```
Comment (jinja2) [inline processor/resourcedetectionprocessor/internal/aws/eks/detector.go:36]: I would recommend to also check for the cluster name as part of the tag key `kubernetes.io/cluster/CLUSTER_NAME`. I think this tag is most likely to be available on older clusters with self-managed ASGs.

File: processor/resourcedetectionprocessor/internal/aws/eks/detector.go [hunk 8]
```diff
@@ -114,3 +129,44 @@ func (e eksDetectorUtils) getConfigMap(ctx context.Context, namespace string, na
...
+	if err != nil {
+		return "", err
+	}
+
+	ec2IMS := imds.NewFromConfig(cfg)
+	regionOutput, err := ec2IMS.GetRegion(context.TODO(), &imds.GetRegionInput{})
+	if err != nil {
+		return "", err
+	}
+	cfg.Region = regionOutput.Region
```
Comment (jinja2) [inline processor/resourcedetectionprocessor/internal/aws/eks/detector.go:144]:
```suggestion
	cfg, err := config.LoadDefaultConfig(context.TODO(), config.WithEC2IMDSRegion())
	if err != nil {
		return "", err
	}
```

Looking at the [default loader in the sdk](https://github.com/aws/aws-sdk-go-v2/blob/v1.21.0/config/config.go#L51), passing in this helper fn should make it query the region from imds.

File: processor/resourcedetectionprocessor/internal/aws/eks/detector.go [hunk 9]
```diff
@@ -114,3 +129,44 @@ func (e eksDetectorUtils) getConfigMap(ctx context.Context, namespace string, na
...
+
+	client := ec2.NewFromConfig(cfg)
+	describeInstances, err := client.DescribeInstances(context.TODO(), &ec2.DescribeInstancesInput{})
```
Comment (jinja2) [inline processor/resourcedetectionprocessor/internal/aws/eks/detector.go:147]: I think this will query all ec2 instances, and we should get the tags for the local instance only. We can get the instance-id from imds and pass the instance id to `DescribeInstancesInput`.

File: processor/resourcedetectionprocessor/internal/aws/eks/detector.go (hunk 7 above)
Comment (crobert-1) [reply processor/resourcedetectionprocessor/internal/aws/eks/detector.go:36]: Thanks, I'll add.

File: processor/resourcedetectionprocessor/internal/aws/eks/detector.go (hunk 8 above)
Comment (crobert-1) [reply processor/resourcedetectionprocessor/internal/aws/eks/detector.go:144]: Tested and it works, I'll update. Thanks!

File: processor/resourcedetectionprocessor/internal/aws/eks/detector.go (hunk 9 above)
Comment (crobert-1) [reply processor/resourcedetectionprocessor/internal/aws/eks/detector.go:147]: Done.

File: processor/resourcedetectionprocessor/internal/aws/eks/detector.go (hunk 2 above)
Comment (mx-psi) [reply processor/resourcedetectionprocessor/internal/aws/eks/detector.go]: I would vote for setting a longer timeout directly (the timeout comes from https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/96d53a20663a0f593565fb84abbad99db3ce88da/processor/resourcedetectionprocessor/config.go#L38-L39C23 )

Comment (crobert-1) [reply processor/resourcedetectionprocessor/internal/aws/eks/detector.go]: Done. I set the default timeout to 11 seconds when there's an EKS detector, which is the cutoff between success and failure in my demo environment for getting the cluster name.

File: processor/resourcedetectionprocessor/internal/aws/eks/metadata.yaml (hunk 1 above)
Comment (jinja2) [reply processor/resourcedetectionprocessor/internal/aws/eks/metadata.yaml:17]: Users might have other processor adding this attribute, and since there is no official cluster name concept in k8s itself, they could've set it to some variation of eks name. I think it would be safer to leave it disabled initially.

File: processor/resourcedetectionprocessor/README.md [hunk 10]
```diff
@@ -318,14 +318,18 @@ processors:
 
     * cloud.provider ("aws")
     * cloud.platform ("aws_eks")
+    * k8s.cluster.name
+
+Note: The kubernetes cluster name is only available when running on EC2 instances, and requires permission to run the `EC2:DescribeInstances` [action](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_DescribeInstances.html).
```
Comment (jinja2) [inline processor/resourcedetectionprocessor/README.md:323]: We can also check if tags are available on the IMDS endpoint first. So if they are exposed on imds, we don't have to make the describe instance call. EC2 instances have a [setting to expose tags through IMDS](https://aws.amazon.com/about-aws/whats-new/2022/01/instance-tags-amazon-ec2-instance-metadata-service/).

Comment (crobert-1) [reply processor/resourcedetectionprocessor/README.md:323]: I know it's possible, but it seemed more complicated to have two routes here. In my test environment `DescribeInstances` was allows, but `DescribeTags` wasn't by default, so this solution was what worked for me. 

Your suggestion would help in the case where permission is granted to run `DescribeTags` but not `DescribeInstances`. Do you think there's some level of likelihood that this would happen? If it is I'm happy to add the check, but I'd prefer to keep it simple if it's not something we expect to run into.

Comment (jinja2) [reply processor/resourcedetectionprocessor/README.md:323]: Right, but this is trying to access tags from the local instance metadata service which shouldn't require an IAM role, just access to the local imds endpoint (the same way we fetched the instance id). But nevermind, the tags are not available on the imds endpoint by default and requires an additional config to be set in the ec2 instances.

**Comment by crobert-1** [comment]
_Sorry for all of the updates here, linting is failing with OOM on the GitHub runners. I'm trying to find the best balance here of getting lint to complete successfully in the least amount of time possible._

File: .golangci.yml [hunk 11]
```diff
@@ -4,7 +4,7 @@ run:
   concurrency: 4
 
   # timeout for analysis, e.g. 30s, 5m, default is 1m
-  timeout: 20m
+  timeout: 100m
```
Comment (atoulme) [inline .golangci.yml:7]: is this still necessary?

File: .golangci.yml [hunk 12]
```diff
@@ -21,6 +21,7 @@ run:
     - local
     - cmd/otelcontribcol
     - cmd/oteltestbedcol
+    - cmd
```
Comment (atoulme) [inline .golangci.yml:24]: can be peeled off into a separate PR

File: Makefile.Common [hunk 13]
```diff
@@ -23,8 +23,8 @@ SRC_ROOT := $(shell git rev-parse --show-toplevel)
 
 # build tags required by any component should be defined as an independent variables and later added to GO_BUILD_TAGS below
 GO_BUILD_TAGS=""
-GOTEST_OPT?= -race -timeout 300s -parallel 4 --tags=$(GO_BUILD_TAGS)
-GOTEST_INTEGRATION_OPT?= -race -timeout 360s -parallel 4
+GOTEST_OPT?= -race -timeout 100m -parallel 4 --tags=$(GO_BUILD_TAGS)
+GOTEST_INTEGRATION_OPT?= -race -timeout 100m -parallel 4
```
Comment (atoulme) [inline Makefile.Common:27]: same, please revert

File: Makefile.Common [hunk 14]
```diff
@@ -174,7 +174,7 @@ fmt: $(GOIMPORTS)
 
 .PHONY: lint
 lint: $(LINT) checklicense misspell
-	$(LINT) run --allow-parallel-runners --verbose --build-tags integration --timeout=30m --path-prefix $(shell basename "$(CURDIR)")
+	GOGC=45 $(LINT) run --allow-parallel-runners --verbose --build-tags integration --timeout=100m --path-prefix $(shell basename "$(CURDIR)")
```
Comment (atoulme) [inline Makefile.Common:177]: please revert

File: .golangci.yml (hunk 11 above)
Comment (crobert-1) [reply .golangci.yml:7]: My apologies, all of the changes you marked were a part of my testing to determine the failure point in the linting process. I'm going to revert as many of these changes as possible, and mark the review as `Ready for Review` when it's ready again.

Comment (atoulme) [reply .golangci.yml:7]: No apologies necessary :) happy to review when it's good to go

**Comment by dmitryax** [comment]
Looks like your linter config change didn't work out. I see `make -C cmd/otelcontribcol lint` in the tests. The linter uses `cmd/otelcontribcol` as root directory

**Comment by crobert-1** [comment]
Superseded by #28649. `v1` of the `aws-sdk-go` library doesn't run into the linting issues, so I've decided to use that instead.
